File: examples/player-python/SettingsModel.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the feather-tk project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl

import json
import weakref

class Model(ftk.Settings):
    """
    This model provides settings that are saved and restored.
    """
    def __init__(self, context):
        docPath = ftk.getUserPath(ftk.UserPath.Documents)
        settingsPath = ftk.getSettingsPath(
            ftk.Path(docPath, "tlRender").get(),
            "player-python.json")
        ftk.Settings.__init__(self, context, settingsPath)
        
        self._contextWeak = weakref.ref(context)

        # Restore file browser settings.
        fileBrowserSystem = context.getSystemByName("ftk::FileBrowserSystem")
        settings = self.getJSON("/FileBrowser")
        if settings[0]:
            fileBrowserSystem.model.options.from_json(settings[1])
        settings = self.getBool("/NativeFileDialog")
        if settings[0]:
            fileBrowserSystem.nativeFileDialog = settings[1]

        # Restore timeline player cache settings.
        cacheOptions = tl.timeline.PlayerCacheOptions()
        settings = self.getJSON("/Cache")
        if settings[0]:
            cacheOptions.from_json(settings[1])
        self._cache = tl.timeline.ObservablePlayerCacheOptions(cacheOptions)

    def __del__(self):
    
        # Save file browse settings.
        fileBrowserSystem = self._contextWeak().getSystemByName("ftk::FileBrowserSystem")
        self.setJSON("/FileBrowser", fileBrowserSystem.model.options.to_json())
        self.setBool("/NativeFileDialog", fileBrowserSystem.nativeFileDialog)
        
        # Save timeline player cache settings.
        self.setJSON("/Cache", self._cache.get().to_json())

    def observeCache(self):
        """
        Observe the timeline player cache settings.
        """
        return self._cache

    def setCache(self, value):
        """
        Set the timeline player cache settings.
        """
        self._cache.setIfChanged(value)

File: lib/tlRender/PlayApp/SettingsModel.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the tlRender project.

#include "SettingsModel.h"

namespace tl
{
    namespace play
    {
        SettingsModel::SettingsModel(
            const std::shared_ptr<ftk::Context>& context,
            const std::filesystem::path& path) :
            Settings(context, path, false)
        {
            // Restore file browser settings.
            ftk::FileBrowserOptions fileBrowserOptions;
            getT("/FileBrowser", fileBrowserOptions);
            auto fileBrowserSystem = context->getSystem<ftk::FileBrowserSystem>();
            fileBrowserSystem->getModel()->setOptions(fileBrowserOptions);
            bool nativeFileDialog = false;
            get("/NativeFileDialog", nativeFileDialog);
            fileBrowserSystem->setNativeFileDialog(nativeFileDialog);
            _fileBrowserSystem = fileBrowserSystem;

            // Restore timeline player cache settings.
            PlayerCacheOptions cache;
            getT("/Cache", cache);
            _cache = ftk::Observable<PlayerCacheOptions>::create(cache);
        }
        
        SettingsModel::~SettingsModel()
        {
            // Save file browser settings.
            if (auto fileBrowserSystem = _fileBrowserSystem.lock())
            {
                setT("/FileBrowser", fileBrowserSystem->getModel()->getOptions());
                set("/NativeFileDialog", fileBrowserSystem->isNativeFileDialog());
            }

            // Save timeline player cache settings.
            setT("/Cache", _cache->get());
        }

        std::shared_ptr<SettingsModel> SettingsModel::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::filesystem::path& path)
        {
            return std::shared_ptr<SettingsModel>(new SettingsModel(context, path));
        }

        const PlayerCacheOptions& SettingsModel::getCache() const
        {
            return _cache->get();
        }

        std::shared_ptr<ftk::IObservable<PlayerCacheOptions> > SettingsModel::observeCache() const
        {
            return _cache;
        }

        void SettingsModel::setCache(const PlayerCacheOptions& value)
        {
            _cache->setIfChanged(value);
        }
    }
}
